xds/src/main/java/com/linecorp/centraldogma/xds/internal/ControlPlanePlugin.java

```diff
@@ -0,0 +1,343 @@
+                                                                     .repos();
+        watchRepository(repositoryManager.get(CLUSTER_REPO), CLUSTER_FILE, Revision.INIT,
+                        (entries, revision) -> clusters(entries, revision, cache));
+        watchRepository(repositoryManager.get(ENDPOINT_REPO), ENDPOINT_FILE, Revision.INIT,
+                        (entries, revision) -> endpoints(entries, revision, cache));
+        watchRepository(repositoryManager.get(LISTENER_REPO), LISTENER_FILE, Revision.INIT,
+                        (entries, revision) -> listeners(entries, revision, cache));
+        watchRepository(repositoryManager.get(ROUTE_REPO), ROUTE_FILE, Revision.INIT,
+                        (entries, revision) -> routes(entries, revision, cache));
+        final V3DiscoveryServer server = new V3DiscoveryServer(new LoggingDiscoveryServerCallbacks(), cache);
+        // xDS, ADS
+        final GrpcService grpcService = GrpcService.builder()
+                                                   .addService(server.getClusterDiscoveryServiceImpl())
+                                                   .addService(server.getEndpointDiscoveryServiceImpl())
+                                                   .addService(server.getListenerDiscoveryServiceImpl())
+                                                   .addService(server.getRouteDiscoveryServiceImpl())
+                                                   .addService(server.getSecretDiscoveryServiceImpl())
+                                                   .addService(server.getAggregatedDiscoveryServiceImpl())
+                                                   .useBlockingTaskExecutor(true)
+                                                   .build();
+        sb.route().requestTimeoutMillis(0).build(grpcService);
+    }
+
+    private void watchRepository(Repository repository, String fileName, Revision revision,
+                                 BiFunction<Collection<Entry<?>>, Revision, Boolean> snapshotConverter) {
+        final CompletableFuture<Revision> future = repository.watch(revision, "/**");
+        future.handleAsync((BiFunction<Revision, Throwable, Void>) (watchedRevision, cause) -> {
+            if (stop) {
+                return null;
+            }
+            if (cause != null) {
+                logger.warn("Unexpected exception is raised while watching {}. Try watching after {} seconds..",
+                            repository, BACKOFF_SECONDS, cause);
+                CONTROL_PLANE_EXECUTOR.schedule(
+                        () -> watchRepository(repository, fileName, revision, snapshotConverter),
+                        BACKOFF_SECONDS, TimeUnit.SECONDS);
+                return null;
+            }
+            final CompletableFuture<Map<String, Entry<?>>> entriesFuture =
+                    repository.find(watchedRevision, "/**/" + fileName);
```

Comment:
Side-note) The bulk read operation may cause high GC pressure when many services use xDS. Streaming or pagination may be necessary later.

Reply:
The implementation will be updated to utilize the customized SnapshotCache, which does not perform a complete refresh for all resources. Please stay tuned for the next PR. 😉